File: send/build-eml.mjs
#!/usr/bin/env node
/**
 * Build a self-contained .eml file from the empeo email template, with the
 * 4 logo/social images EMBEDDED inline (CID) so they always display — even
 * before "Download pictures" and even offline.
 *
 * Double-click the resulting .eml in Outlook → it opens as a NEW editable
 * email (thanks to the `X-Unsent: 1` header) → fill in "To" → replace the
 * name → hit Send.
 *
 * Usage:
 *   node build-eml.mjs ["<Full Name>"]
 *   # default keeps the {{FULL_NAME}} placeholder so you replace it in Outlook
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, '..');

const SUBJECT = process.env.EMAIL_SUBJECT || 'คำขอลบบัญชีของคุณได้รับการอนุมัติ';
const SENDER = process.env.MS_SENDER || 'parisa.a@gofive.co.th';
const fullName = process.argv[2] || '{{FULL_NAME}}';
// Which template to build (default = gray version). e.g. TEMPLATE=empeo-account-deletion-white.html
const templateName = process.env.TEMPLATE || 'empeo-account-deletion.html';

// ---------- Load template, swap placeholders + image URLs → cid: ----------
let html = fs.readFileSync(
  path.join(root, 'email-templates', templateName),
  'utf8',
)
  .replaceAll('{{FULL_NAME}}', fullName)
  // Optional per-template placeholders (default: keep so you can edit in Outlook)
  .replaceAll('{{PAY_PERIOD}}', process.env.PAY_PERIOD || '{{PAY_PERIOD}}')
  .replaceAll('{{DOWNLOAD_URL}}', process.env.DOWNLOAD_URL || '{{DOWNLOAD_URL}}');

// jsDelivr URL (…/assets/<name>.png)  →  cid:<name>
html = html.replace(
  /https:\/\/cdn\.jsdelivr\.net\/gh\/[^"']*\/assets\/([\w-]+)\.png/g,
  (_, name) => `cid:${name}`,
);

// ---------- Helpers ----------
const wrap76 = (b64) => b64.match(/.{1,76}/g).join('\r\n');
const encWord = (s) => `=?UTF-8?B?${Buffer.from(s, 'utf8').toString('base64')}?=`;

// Embed exactly the images this template references (via cid:), in first-seen order.
const imageNames = [...new Set([...html.matchAll(/cid:([\w-]+)/g)].map((m) => m[1]))];
const images = imageNames.map((name) => ({
  name,
  cid: name,
  b64: wrap76(fs.readFileSync(path.join(root, 'email-templates', 'assets', `${name}.png`)).toString('base64')),
}));

// ---------- Assemble MIME (multipart/related) ----------
const BOUNDARY = 'empeo_boundary_5f3a1c9b';
const CRLF = '\r\n';

let eml = '';
eml += `From: empeo <${SENDER}>` + CRLF;
eml += `To: ` + CRLF;
eml += `Subject: ${encWord(SUBJECT)}` + CRLF;
eml += `X-Unsent: 1` + CRLF; // makes Outlook open it as a new, sendable draft
eml += `MIME-Version: 1.0` + CRLF;
eml += `Content-Type: multipart/related; boundary="${BOUNDARY}"` + CRLF + CRLF;

// HTML part (base64 to safely carry Thai + long lines)
eml += `--${BOUNDARY}` + CRLF;
eml += `Content-Type: text/html; charset="UTF-8"` + CRLF;
eml += `Content-Transfer-Encoding: base64` + CRLF + CRLF;
eml += wrap76(Buffer.from(html, 'utf8').toString('base64')) + CRLF + CRLF;

// Inline image parts
for (const img of images) {
  eml += `--${BOUNDARY}` + CRLF;
  eml += `Content-Type: image/png; name="${img.name}.png"` + CRLF;
  eml += `Content-Transfer-Encoding: base64` + CRLF;
  eml += `Content-ID: <${img.cid}>` + CRLF;
  eml += `Content-Disposition: inline; filename="${img.name}.png"` + CRLF + CRLF;
  eml += img.b64 + CRLF + CRLF;
}
eml += `--${BOUNDARY}--` + CRLF;

// ---------- Write ----------
const outPath = path.join(__dirname, templateName.replace(/\.html$/, '.eml'));
fs.writeFileSync(outPath, eml);
console.log(`✅ Wrote ${path.relative(root, outPath)} (${(eml.length / 1024).toFixed(0)} KB, images embedded)`);
console.log(`   Name in body: ${fullName === '{{FULL_NAME}}' ? '{{FULL_NAME}} (replace in Outlook)' : fullName}`);

File: emails/welcome-onboarding/tools/build_eml.py
#!/usr/bin/env python3
"""Assemble welcome-onboarding.html + images/ into a single .eml file.

Every image referenced as src="cid:NAME" is attached as images/NAME.png with a
matching Content-ID, so the message renders offline with no remote hosting.

    python3 tools/build_eml.py [-o out.eml]
"""

import argparse
import base64
import os
import re

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.normpath(os.path.join(HERE, '..'))

HTML = os.path.join(ROOT, 'welcome-onboarding.html')
IMAGES = os.path.join(ROOT, 'images')

FROM = 'empeo <parisa.a@gofive.co.th>'
SUBJECT = 'ยินดีต้อนรับสู่ทีม เข้าใช้งาน empeo ได้เลย'
BOUNDARY = 'empeo_boundary_5f3a1c9b'


def b64_lines(raw):
    """Base64 wrapped at 76 characters, as MIME wants it."""
    enc = base64.b64encode(raw).decode('ascii')
    return '\r\n'.join(enc[i:i + 76] for i in range(0, len(enc), 76))


def build():
    with open(HTML, 'rb') as fh:
        html = fh.read()

    cids = []
    for cid in re.findall(rb'src="cid:([^"]+)"', html):
        cid = cid.decode('ascii')
        if cid not in cids:
            cids.append(cid)

    parts = [
        'From: %s' % FROM,
        'To: ',
        'Subject: =?UTF-8?B?%s?=' % base64.b64encode(SUBJECT.encode('utf-8')).decode('ascii'),
        'X-Unsent: 1',
        'MIME-Version: 1.0',
        'Content-Type: multipart/related; boundary="%s"' % BOUNDARY,
        '',
        '--%s' % BOUNDARY,
        'Content-Type: text/html; charset="UTF-8"',
        'Content-Transfer-Encoding: base64',
        '',
        b64_lines(html),
        '',
    ]

    for cid in cids:
        path = os.path.join(IMAGES, cid + '.png')
        if not os.path.exists(path):
            raise SystemExit('missing image for cid:%s (expected %s)' % (cid, path))
        with open(path, 'rb') as fh:
            raw = fh.read()
        parts += [
            '--%s' % BOUNDARY,
            'Content-Type: image/png; name="%s.png"' % cid,
            'Content-Transfer-Encoding: base64',
            'Content-ID: <%s>' % cid,
            'Content-Disposition: inline; filename="%s.png"' % cid,
            '',
            b64_lines(raw),
            '',
        ]

    parts += ['--%s--' % BOUNDARY, '']
    return '\r\n'.join(parts), cids


if __name__ == '__main__':
    ap = argparse.ArgumentParser()
    ap.add_argument('-o', '--out', default=os.path.join(ROOT, 'empeo-welcome-onboarding.eml'))
    args = ap.parse_args()

    eml, cids = build()
    with open(args.out, 'w', encoding='utf-8', newline='') as fh:
        fh.write(eml)
    print('wrote %s (%d bytes, %d inline images: %s)'
          % (args.out, len(eml), len(cids), ', '.join(cids)))
